compiler/src/dotty/tools/dotc/transform/patmat/Space.scala

```diff
@@ -628,6 +628,13 @@ class SpaceEngine(using Context) extends SpaceLogic {
         Typ(ConstantType(Constant(())), true) :: Nil
       case tp if tp.classSymbol.isAllOf(JavaEnumTrait) =>
         tp.classSymbol.children.map(sym => Typ(sym.termRef, true))
+
+      case tp @ AppliedType(tycon, targs) if tp.classSymbol.children.isEmpty && canDecompose(tycon) =>
+        rec(tycon, Nil).map(typ => Typ(tp.derivedAppliedType(typ.tp, targs)))
```

Comment:
It's not obvious that it's OK to apply the type arguments of a parent type to child types. Maybe add a comment given that the condition `tp.classSymbol.children.isEmpty && canDecompose(tycon)` is uncommon.